### PR TITLE
Added UI Render for UI Grid logic

### PR DESCRIPTION
**Pull request by pavel-trafimuk**

### CODE REVIEW

**Comment by pavel-trafimuk** [comment]
@bigsnickers another one

File: Sources/UIGrid/UIGridButton.swift [hunk 1]
```diff
@@ -5,20 +5,13 @@
...
-//
-
 public struct UIGridButton: Codable, Equatable {
 
     public let columns: UInt
     public let rows: UInt
 
     public var size: UIGridSize {
-        .init(columns: columns, rows: rows)
+        .init(width: columns, height: rows)
```
Comment (bigsnickers) [inline Sources/UIGrid/UIGridButton.swift]: This one will compile slowly. Use UIGridSize directly

File: Sources/UIGrid/UIGridSize.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        self.height = height
+    }
+    
+    public let width: UInt
+    public let height: UInt
+    
+    public var zero: UIGridSize {
+        .init(width: 0, height: 0)
+    }
+        public var isZero: Bool {
```
Comment (bigsnickers) [inline Sources/UIGrid/UIGridSize.swift]: indentation

File: Sources/UIGrid/UIGridSize.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        width == 0 && height == 0
```
Comment (bigsnickers) [inline Sources/UIGrid/UIGridSize.swift:22]: maybe || ?

File: Sources/UIGridView/BotKeyboardView.swift [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        collectionView.isAccessibilityElement = false
+        addSubview(collectionView)
+        self.collectionView = collectionView
+        
+        // first sizerToFit will automatically rotate collectionView to needed size (and reload it)
+        sizeToFit()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (bigsnickers) [inline Sources/UIGridView/BotKeyboardView.swift]: preconditionFailure

File: Sources/UIGridView/BotKeyboardView.swift [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        }
+        else {
+            collectionView.frame = collectionViewFrame
+        }
+
+        // TODO: seems like it will flick!
+        if viewModel.grid.isDefaultHeight != true {
+            collectionViewFrame.size.height = min(collectionView.contentSize.height, defaultKeyboardSize.height - collectionViewFrame.origin.y)
+            collectionViewFrame.size.height += superview?.safeAreaInsets.bottom ?? 0.0
+            collectionView.frame = collectionViewFrame
```
Comment (bigsnickers) [inline Sources/UIGridView/BotKeyboardView.swift:209]: why can't we simply use autolayout here?

File: Sources/UIGridView/GradientView.swift [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        switch location {
+        case .top:
+            gradientLayer.colors = [startColor, endColor]
+        case .bottom:
+            gradientLayer.colors = [endColor, startColor]
+        }
+        gradientLayer.locations = [NSNumber(value: 0.0), NSNumber(value: 1.0)]
+    }
+
+    override var bounds: CGRect {
```
Comment (bigsnickers) [inline Sources/UIGridView/GradientView.swift]: just override layoutSubviews()

File: Sources/UIGridView/UIGridButtonView.swift [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+    
+    override public var isSelected: Bool {
+        didSet {
+            updateForPressedState()
+        }
+    }
+    
+    override public var isHighlighted: Bool {
+        didSet {
+            updateForPressedState()
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridButtonView.swift]: if isHighlighted != oldValue

File: Sources/UIGridView/UIGridButtonViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    
+    /// not force will do it after end of current
+    /// TODO:     self.backgroundGIFView.loopCompletionBlock = @weakselfnotnil(^(NSUInteger loop)) {
+//    [self.backgroundGIFView stopAnimating]
+//} @weakselfend
+
+    func stopAnimation()
+}
+
+public protocol UIGridMediaProvider {
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridButtonViewModel.swift]: : AnyObject if implemented by classes

File: Sources/UIGridView/UIGridCollectionLayout.swift [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    public var gridGroupSize: UIGridSize {
+        didSet {
+            if gridGroupSize != oldValue {
+                invalidateLayout()
+            }
+        }
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridCollectionLayout.swift]: preconditionFailure

File: Sources/UIGridView/UIGridCollectionLayout.swift [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+            let storage = try UIGridLayoutStorageGenerator().generateStorage(with: loadedValues,
+                                                                       limitOfGroupsInOneRow: limitOfGroupsInOneRow,
+                                                                       gridGroupSize: gridGroupSize,
+                                                                       reverseOrderForGroups: false,
+                                                                       flexibleRowsLayout: true) // TODO: check values!
+            cachedStorages[limitOfGroupsInOneRow * gridGroupSize.width] = storage
+            activeGridStorage = storage
+        }
+        catch {
+            print("Error: \(error)")
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridCollectionLayout.swift]: replace with the logger

File: Sources/UIGridView/UIGridLayoutStorage.swift [hunk 11]
```diff
@@ -0,0 +1,265 @@
+//
+//  Created by Pavel Trafimuk
+//  Copyright © 2023 Viber Media Sarl. All rights reserved.
+//
+
+import Foundation
+import UIGrid
+
+//public struct UIGridFrameIterator: IteratorProtocol {
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridLayoutStorage.swift:9]: TODO

File: Sources/UIGridView/UIGridViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+                publicAccountId: String? = nil,
+                actionHandler: UIGridViewActionHandling?,
+                defaultBackgroundColor: UIColor,
+                mediaProvider: UIGridMediaProvider?) {
+        self.grid = grid
+        self.publicAccountId = publicAccountId
+        self.actionHandler = actionHandler
+        self.mediaProvider = mediaProvider
+        
+        //    NSError *unError = nil;
```
Comment (bigsnickers) [inline Sources/UIGridView/UIGridViewModel.swift:123]: do something with that

File: Tests/UIGridTests/UIGridTests.swift [hunk 13]
```diff
@@ -543,15 +543,15 @@ final class UIGridTests: XCTestCase {
...
+    func testRichMessagesDefaultgridGroupSize_0() throws {
         let vm = try UIGrid.load(from: "richMessagesJSONExample")
         XCTAssert(vm.buttons.count == 5, "Invalid count of buttons in rich message")
         XCTAssert(vm.buttonsGroupColumns == 6)
         XCTAssert(vm.buttonsGroupRows == 7)
     }
     
     
-    func testRichMessagesDefaultGroupSize_1() throws {
+    func testRichMessagesDefaultgridGroupSize_1() throws {
```
Comment (bigsnickers) [inline Tests/UIGridTests/UIGridTests.swift]: typo

File: Tests/UIGridViewTests/UIGridLayoutStorageGeneratorTests.swift [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+    override func setUpWithError() throws {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    }
+
+    func testZeroWidth() throws {
+//        NSDictionary *elements = @{};
```
Comment (bigsnickers) [inline Tests/UIGridViewTests/UIGridLayoutStorageGeneratorTests.swift:28]: TODO

**Comment by bigsnickers** [comment]
Reviewed

File: Sources/UIGrid/UIGridSize.swift (hunk 3 above)
Comment (pavel-trafimuk) [reply Sources/UIGrid/UIGridSize.swift:22]: not - &&

File: Sources/UIGridView/BotKeyboardView.swift (hunk 5 above)
Comment (pavel-trafimuk) [reply Sources/UIGridView/BotKeyboardView.swift:209]: frame in half time depends on json values, so jumps between auto layout and manual frame is not good idea

File: Sources/UIGridView/GradientView.swift (hunk 6 above)
Comment (pavel-trafimuk) [reply Sources/UIGridView/GradientView.swift]: was copied from previous realisation. fixed

**Pull request merged**